VariantShredding.md

```diff
@@ -25,290 +25,316 @@
 
-Shredding allows a query engine to reap the full benefits of Parquet's columnar representation, such as more compact data encoding, min/max statistics for data skipping, and I/O and CPU savings from pruning unnecessary fields not accessed by a query (including the non-shredded Variant binary data).
-Without shredding, any query that accesses a Variant column must fetch all bytes of the full binary buffer.
-With shredding, we can get nearly equivalent performance as in a relational (scalar) data model.
+## Variant Metadata
 
-For example, `select variant_get(variant_col, ‘$.field1.inner_field2’, ‘string’) from tbl` only needs to access `inner_field2`, and the file scan could avoid fetching the rest of the Variant value if this field was shredded into a separate column in the Parquet schema.
-Similarly, for the query `select * from tbl where variant_get(variant_col, ‘$.id’, ‘integer’) = 123`, the scan could first decode the shredded `id` column, and only fetch/decode the full Variant value for rows that pass the filter.
+Variant metadata is stored in the top-level Variant group in a binary `metadata` column regardless of whether the Variant value is shredded.
 
-# Parquet Example
+All `value` columns within the Variant must use the same `metadata`.
+All field names of a Variant, whether shredded or not, must be present in the metadata.
 
-Consider the following Parquet schema together with how Variant values might be mapped to it.
-Notice that we represent each shredded field in `object` as a group of two fields, `typed_value` and `variant_value`.
-We extract all homogenous data items of a certain path into `typed_value`, and set aside incompatible data items in `variant_value`.
-Intuitively, incompatibilities within the same path may occur because we store the shredding schema per Parquet file, and each file can contain several row groups.
-Selecting a type for each field that is acceptable for all rows would be impractical because it would require buffering the contents of an entire file before writing.
+## Value Shredding
 
-Typically, the expectation is that `variant_value` exists at every level as an option, along with one of `object`, `array` or `typed_value`.
-If the actual Variant value contains a type that does not match the provided schema, it is stored in `variant_value`.
-An `variant_value` may also be populated if an object can be partially represented: any fields that are present in the schema must be written to those fields, and any missing fields are written to `variant_value`.
-
-The `metadata` column is unchanged from its unshredded representation, and may be referenced in `variant_value` fields in the shredded data.
+Variant values are stored in Parquet fields named `value`.
+Each `value` field may have an associated shredded field named `typed_value` that stores the value when it matches a specific type.
```

Comment:
I think we should be explicit that in this version of the spec, there can be exactly one "typed_value" column?

Reply:
I wonder if it pays to name fields using something like hungarian notation to allow for easy evolution if we want to support multiple types in the future (I don't feel strongly about this).

Reply:
Isn't this clear? It says "an associated shredded field" with a specific name. I'm not sure how that could be interpreted to allow multiple fields in this version of the spec.

Reply:
```suggestion
Each `value` field may have a single associated shredded field named `typed_value` that stores the value when it matches a specific type.
```
I think you are right and I might have been reading to quickly, at least for me this probably would not have been something I missed (we can close this either way).